src/wwwroot/app/directives/candidate-vacancy-info/candidate-vacancy-info.js

```diff
@@ -29,179 +35,229 @@ export default class CandidateVacancyInfoDirective {
-      let deffered = $q.defer();
-      getCurrentAndSelectedStages(stagesObject).then((stages) => {
-         return goToStage(stagesObject, stages.currentStage, stages.selectedStage);
-      }).then((selectedStage) => {
-         if (selectedStage) {
-            deffered.resolve(selectedStage);
-            notifyCurrentStageChangingForward();
-         }
-         deffered.resolve();
-      }).catch((currentStage) => {
-         deffered.reject(currentStage);
+      return map(rejectStages, (extStage) => {
+         return {
+            stage: extStage,
+            vsi: find(entityStageObject.vacancyStageInfos, { stageId: extStage.stage.id })
+         };
       });
-      return deffered.promise;
-   }
-
-   function getCurrentAndSelectedStages(stagesObject) {
-      let deffer = $q.defer();
-      let vacancyStages = vm.vacancyStages ? vm.vacancyStages : stagesObject.stageFlow;
-      let currentStage = find(vacancyStages, { id: stagesObject.currentStageId });
-      let selectedStage = find(vacancyStages, { id: stagesObject.selectedStageId });
-      deffer.resolve({currentStage, selectedStage});
-      return deffer.promise;
-   };
-
-   function goToStage (stagesObject, currentStage, selectedStage) {
-      let deffered = $q.defer();
-      if (_isSelectedIsNextToCurrent(selectedStage, currentStage)) {
-         handleStagePassing(stagesObject, currentStage).then((actionResultObject) => {
-            return _createOrUpdateVSIWith(stagesObject, currentStage, actionResultObject.comment, actionResultObject.isPassed);  // eslint-disable-line max-len
-         }).then(() => {
-            return _createOrUpdateVSIWith(stagesObject, selectedStage, null, false);
-         }).then(() => {
-            return setCurrentStageBySelected(stagesObject, selectedStage);
-         }).then(() => {
-            deffered.resolve(selectedStage);
-         }).catch(() => {
-            deffered.reject(currentStage);
-         });
-      } else if (selectedStage.id > currentStage.id) {
-         _findStagesAndRecomposeBy(stagesObject).then((recomposed) => {
-            return showManyCommentDialog(recomposed);
-         }).then((dialogTransferObject) => {
-            forEach(dialogTransferObject.stagesToPass, (stageObj) => {
-               _createOrUpdateVSIWith(stagesObject, stageObj.stage, stageObj.comment, stageObj.isPassed);
-            });
-            return setCurrentStageBySelected(stagesObject, selectedStage);
-         }).then(() => {
-            deffered.resolve(selectedStage);
-         })
-         .catch(() => {
-            deffered.reject(currentStage);
-         });
-      } else {
-         deffered.resolve();
-      }
-      return deffered.promise;
    }
 
-   function _isSelectedIsNextToCurrent(selectedStage, currentStage) {
-      return currentStage.order + 1 === selectedStage.order;
+   function keepHireVSI(entityStageObject, vacancyStages) {
+      let hireStage = head(filter(vacancyStages, (extStage) => {
```

Comment:
you can create named function for stage cheking predicate in filter and curry it. in thei case your code can be next:

function isExtStageHasType(stageType. stage) {
  return  stage.stageType === stageType;
}
//...
let _isExtStageHasType = curry(isExtStageHasType, 2);
//...
head(filter(vacancyStages, _isExtStageHasType(STAGE_TYPES.HireStage));